pkg/controller/populators/clone-populator.go

```diff
@@ -0,0 +1,379 @@
+	log.V(1).Info("reconciling Clone Source PVC")
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(ctx, req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// replace this eventually
+	if pvc.Spec.DataSourceRef != nil &&
+		pvc.Spec.DataSourceRef.Namespace != nil &&
+		*pvc.Spec.DataSourceRef.Namespace != pvc.Namespace {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, fmt.Errorf("cross namespace datasource not supported yet"))
+	}
+
+	hasFinalizer := cc.HasFinalizer(pvc, cloneFinalizer)
+	isBound := pvc.Spec.VolumeName != ""
+	isDeleted := !pvc.DeletionTimestamp.IsZero()
+
+	log.V(1).Info("pvc state", "hasFinalizer", hasFinalizer, "isBound", isBound, "isDeleted", isDeleted)
+
+	if !isDeleted && !isBound {
+		return r.reconcilePending(ctx, log, pvc)
+	}
+
+	if (isBound || isDeleted) && hasFinalizer {
+		if isBound && !isClonePhaseSucceeded(pvc) {
+			log.V(1).Info("setting phase to Succeeded")
+			return reconcile.Result{}, r.updateClonePhaseSucceeded(ctx, pvc)
+		}
+
+		return r.reconcileDone(ctx, log, pvc)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ClonePopulatorReconciler) reconcilePending(ctx context.Context, log logr.Logger, pvc *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
+	ready, _, err := claimReadyForPopulation(ctx, r.client, pvc)
+	if err != nil {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, err)
+	}
+
+	if !ready {
+		log.V(3).Info("claim not ready for population, exiting")
+		return reconcile.Result{}, r.updateClonePhasePending(ctx, pvc)
+	}
+
+	vcs, err := clone.GetVolumeCloneSource(ctx, r.client, pvc)
+	if err != nil {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, err)
+	}
+
+	if vcs == nil {
+		log.V(3).Info("dataSourceRef does not exist, exiting")
+		return reconcile.Result{}, r.updateClonePhasePending(ctx, pvc)
+	}
+
+	cs := getSavedCloneStrategy(pvc)
+	if cs == nil {
+		args := &clone.ChooseStrategyArgs{
+			Log:         log,
+			TargetClaim: pvc,
+			DataSource:  vcs,
+		}
+
+		cs, err = r.planner.ChooseStrategy(ctx, args)
+		if err != nil {
+			return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, err)
+		}
+
+		if cs == nil {
+			log.V(3).Info("unable to choose clone strategy now")
+			// TODO create index/watch to deal with this
+			return reconcile.Result{RequeueAfter: 5 * time.Second}, r.updateClonePhasePending(ctx, pvc)
+		}
+	}
+
+	updated, err := r.initTargetClaim(ctx, pvc, vcs, *cs)
+	if err != nil {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, err)
+	}
+
+	if updated {
+		log.V(3).Info("initialized target, returning")
+		// phase will be set to pending by initTargetClaim if unset
+		return reconcile.Result{}, nil
+	}
+
+	args := &clone.PlanArgs{
+		Log:         log,
+		TargetClaim: pvc,
+		DataSource:  vcs,
+		Strategy:    *cs,
+	}
+
+	phases, err := r.planner.Plan(ctx, args)
```

Comment:
Really like this design, but I can't help but wonder how other users/contributors will figure out what's happening here.  I think this would benefit by having more comments or even by dividing reconcilePending into smaller functions.

Reply:
I'll think about how it can be clearer.